homeassistant/components/template/template_entity.py

```diff
@@ -430,3 +445,220 @@ async def async_added_to_hass(self) -> None:
     async def async_update(self) -> None:
         """Call for forced update."""
         self._async_update()
+
+
+class TemplateRestoreEntity(RestoreEntity, TemplateEntity):
+    """Template Entity that restores data."""
+
+    def __init__(self, *args, **kwargs) -> None:
+        """Template Restore Entity init."""
+        super().__init__(*args, **kwargs)
+        self._restore: bool = False
+        self._save_state: bool = True
+        self._additional_data: list[str] = []
+
+    def add_template_attribute(
+        self, attribute: str, template: Template, *args, **kwargs
+    ):
+        """Store attribute to allow restore and setup template."""
+
+        super().add_template_attribute(attribute, template, *args, **kwargs)
+        if not template.is_static:
+            self.add_additional_data(attribute)
+
+    @callback
+    def _handle_results(self, *args, **kwargs) -> None:
+        # Do not save state if handle result fails.
+        self._save_state = False
+        super()._handle_results(*args, **kwargs)
+        self._save_state = True
+
+    @property
+    def restore(self) -> bool:
+        """Retrieve restore."""
+        return self._restore or False
+
+    @restore.setter
+    def restore(self, restore: bool) -> None:
+        """Set restore."""
+        self._restore = restore
```

Comment:
Why do we need to hide self._restore behind getter+setter, who is calling these?

Reply:
That problem should be fixed by making sure classes extending `TemplateRestoreEntity` call `super().__init__()` or `TemplateRestoreEntity.__init__`.

Reply:
This is not there anymore as it is now all part of TemplateEntity and TriggerEntity themselves and thus always get called.